monitoring/asset_collector.go

```diff
@@ -0,0 +1,107 @@
+package monitoring
+
+import (
+	"context"
+	"sync"
+
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const (
+	assetCollectorName = "asset"
+
+	numAssetsMintedMetric = "num_assets_minted"
+)
+
+type assetCollector struct {
+	collectMx sync.Mutex
+
+	cfg *PrometheusConfig
+
+	numAssetsMinted prometheus.Gauge
+}
+
+func newAssetCollector(cfg *PrometheusConfig) *assetCollector {
+	return &assetCollector{
+		cfg: cfg,
+		numAssetsMinted: prometheus.NewGauge(
+			prometheus.GaugeOpts{
+				Name: numAssetsMintedMetric,
+				Help: "Total number of assets minted",
+			}),
+	}
+}
+
+func (a *assetCollector) Name() string {
+	return assetCollectorName
+}
+
+// Describe sends the super-set of all possible descriptors of metrics
+// collected by this Collector to the provided channel and returns once the
+// last descriptor has been sent.
+//
+// NOTE: Part of the prometheus.Collector interface.
+func (a *assetCollector) Describe(ch chan<- *prometheus.Desc) {
+	log.Infof("Asset Collector Describe()")
+	a.collectMx.Lock()
+	defer a.collectMx.Unlock()
+
+	a.numAssetsMinted.Describe(ch)
+}
+
+// Collect is called by the Prometheus registry when collecting metrics.
+//
+// NOTE: Part of the prometheus.Collector interface.
+func (a *assetCollector) Collect(ch chan<- prometheus.Metric) {
+	log.Infof("Asset Collector Collect()")
+	a.collectMx.Lock()
+	defer a.collectMx.Unlock()
+
+	if a.cfg == nil {
+		log.Error("cfg is nil")
+		return
+	}
+
+	if a.cfg.UniverseStats == nil {
+		log.Error("UniverseStats is nil")
+		return
+	}
+
+	ctx, cancel := context.WithTimeout(context.Background(), dbTimeout)
+	defer cancel()
+
+	universeStats, err := a.cfg.UniverseStats.AggregateSyncStats(ctx)
```

Comment:
Is there a need at all for such collector? Couldn't we just use SQL for the dash and create the data on demand for the requested time interval?

Reply:
> and create the data on demand for the requested time interval?

How would you gain access to past states?

Prometheus scrapes on a fixed interval, who would do the scraping with manual SQL on the dash?

Reply:
Looking at the query here it makes sense to scrape as this doesn't look very expensive and provides a better user experience as users don't need to write complicated SQL queries.

The alternative would be to use the `event_time` in the `universe_events` table as the time source and create a query generating the same stats bucketed by time intervals. (This could be accomplished using windowing and the `PARTITION BY` clause). Bucketing is useful to provide a level of detail in Grafana.